ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -514,21 +514,31 @@ func (r *RayServiceReconciler) reconcileRayCluster(ctx context.Context, rayServi
 	if shouldUpdateCluster(rayServiceInstance, activeRayCluster, true) {
 		// TODO(kevin85421): We should not reconstruct the cluster to update it. This will cause issues if autoscaler is enabled.
 		logger.Info("Updating the active RayCluster instance", "clusterName", activeRayCluster.Name)
-		if activeRayCluster, err = constructRayClusterForRayService(rayServiceInstance, activeRayCluster.Name, r.Scheme); err != nil {
+		goalCluster, err := constructRayClusterForRayService(rayServiceInstance, activeRayCluster.Name, r.Scheme)
+		if err != nil {
 			return nil, nil, err
 		}
-		err = r.updateRayClusterInstance(ctx, rayServiceInstance, activeRayCluster)
-		return activeRayCluster, pendingRayCluster, err
+		updatedActiveCluster := updateRayClusterInstance(ctx, activeRayCluster, goalCluster)
+		if err = r.Update(ctx, updatedActiveCluster); err != nil {
+			r.Recorder.Eventf(rayServiceInstance, corev1.EventTypeWarning, string(utils.FailedToUpdateRayCluster), "Failed to update the active RayCluster %s/%s: %v", activeRayCluster.Namespace, activeRayCluster.Name, err)
+		}
+		r.Recorder.Eventf(rayServiceInstance, corev1.EventTypeNormal, string(utils.UpdatedRayCluster), "Updated the active RayCluster %s/%s", activeRayCluster.Namespace, activeRayCluster.Name)
+		return updatedActiveCluster, pendingRayCluster, err
 	}
 
 	if shouldUpdateCluster(rayServiceInstance, pendingRayCluster, false) {
 		// TODO(kevin85421): We should not reconstruct the cluster to update it. This will cause issues if autoscaler is enabled.
 		logger.Info("Updating the pending RayCluster instance", "clusterName", pendingRayCluster.Name)
-		if pendingRayCluster, err = constructRayClusterForRayService(rayServiceInstance, pendingRayCluster.Name, r.Scheme); err != nil {
+		goalCluster, err := constructRayClusterForRayService(rayServiceInstance, pendingRayCluster.Name, r.Scheme)
+		if err != nil {
 			return nil, nil, err
 		}
-		err = r.updateRayClusterInstance(ctx, rayServiceInstance, pendingRayCluster)
-		return activeRayCluster, pendingRayCluster, err
+		updatedPendingCluster := updateRayClusterInstance(ctx, pendingRayCluster, goalCluster)
+		if err = r.Update(ctx, updatedPendingCluster); err != nil {
```

Comment:
I think `r.Update(ctx, updatedPendingCluster)` reads confusing. We probably don't need `updateRayClusterInstance` anymore.


```suggestion
		pendingRayCluster.Spec = goalCluster.Spec
		pendingRayCluster.Labels = goalCluster.Labels
		pendingRayCluster.Annotations = goalCluster.Annotations
		if err = r.Update(ctx, pendingRayCluster); err != nil {
```

Reply:
It's better to keep `updateRayClusterInstance` to avoid possible inconsistency between updating active cluster and updating pending cluster.

Maybe I can rename the function to `modifyRayCluster`, and update the CR in-place.

Reply:
Updated. We can sync offline if it is still hard to read.

Reply:
No need. It looks good now.